src/hp/netcheck.rs

```diff
@@ -1106,85 +1065,298 @@ impl ProbeReport {
+    /// The sender side of the messages channel.
+    ///
+    /// This allows creating new [`ActorAddr`]s from the actor.
+    sender: mpsc::Sender<ActorMessage>,
+    /// A collection of previously generated reports.
+    ///
+    /// Sometimes it is useful to look at past reports to decide what to do.
     reports: Reports,
+
+    // Actor configuration.
     /// Whether the client should try to reach things other than localhost.
     ///
     /// This is set to true in tests to avoid probing the local LAN's router, etc.
     skip_external_network: bool,
-    /// Whether to run port mapping queries.
+    /// The port mapper client, if those are requested.
+    ///
+    /// The port mapper is responsible for talking to routers via UPnP and the like to try
+    /// and open ports.
     port_mapper: Option<portmapper::Client>,
-    got_hair_stun: broadcast::Sender<SocketAddr>,
     dns_resolver: TokioAsyncResolver,
+
+    // Actor state.
+    got_hair_stun: broadcast::Sender<SocketAddr>,
+    /// Information about the currently in-flight STUN requests.
+    ///
+    /// This is used to complete the STUN probe when receiving STUN packets.
+    in_flight_stun_requests: HashMap<stun::TransactionId, Inflight>,
+    /// The response channel if there is a check running.
+    ///
+    /// There can only ever be one check running at a time.  If it is running the response
+    /// channel is stored here.
+    current_check_run: Option<oneshot::Sender<Result<Arc<Report>>>>,
 }
 
 impl Actor {
-    /// Performs a single netcheck run.
+    /// Creates a new actor.
     ///
-    /// See [`Client::get_report`] for the arguments.
-    #[instrument(name = "get_report", skip_all)]
-    async fn run(
+    /// This does not start the actor, see [`Actor::main`] for this.  You should not
+    /// normally create this directly but rather create a [`Client`].
+    fn new(port_mapper: Option<portmapper::Client>) -> Result<Self> {
+        // TODO: consider an instrumented flume channel so we have metrics.
+        let (sender, receiver) = mpsc::channel(32);
+        let (got_hair_stun, _) = broadcast::channel(1);
+        let dns_resolver = TokioAsyncResolver::tokio_from_system_conf()
+            .context("failed to create DNS resolver")?;
+        Ok(Self {
+            receiver,
+            sender,
+            reports: Default::default(),
+            skip_external_network: false,
+            port_mapper,
+            got_hair_stun,
+            dns_resolver,
+            in_flight_stun_requests: Default::default(),
+            current_check_run: None,
+        })
+    }
+
+    /// Returns the channel to send messages to the actor.
+    fn addr(&self) -> ActorAddr {
+        ActorAddr {
+            sender: self.sender.clone(),
+        }
+    }
+
+    /// Run the actor.
+    ///
+    /// It will now run and handle messages.  Once the connected [`Client`] (including all
+    /// its clones) is dropped this will terminate.
+    async fn main(&mut self) {
```

Comment:
can we call this `run` for consistency with the other methods in the codebase?

Reply:
sure, i had both functions for a while so needed a new name for the new func as i was converting... then didn't move it back